metrics/src/main/java/org/wildfly/extension/metrics/api/WildFlyMetric.java

```diff
@@ -0,0 +1,84 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.extension.metrics.api;
+
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.FAILURE_DESCRIPTION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.NAME;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.OP_ADDR;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.READ_ATTRIBUTE_OPERATION;
+import static org.jboss.as.controller.descriptions.ModelDescriptionConstants.RESULT;
+import static org.wildfly.extension.metrics._private.MetricsLogger.LOGGER;
+
+import java.util.OptionalDouble;
+
+import org.jboss.as.controller.LocalModelControllerClient;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.as.controller.descriptions.ModelDescriptionConstants;
+import org.jboss.dmr.ModelNode;
+
+public class WildFlyMetric implements Metric {
+
+    private LocalModelControllerClient modelControllerClient;
+    private final PathAddress address;
+    private final String attributeName;
+
+    public WildFlyMetric(LocalModelControllerClient modelControllerClient, PathAddress address, String attributeName) {
+        this.modelControllerClient = modelControllerClient;
+        this.address = address;
+        this.attributeName = attributeName;
+    }
+
+    @Override
+    public OptionalDouble getValue() {
+        ModelNode result = readAttributeValue(address, attributeName);
+        if (result.isDefined()) {
+            try {
+                return OptionalDouble.of(result.asDouble());
+            } catch (Exception e) {
+                throw LOGGER.unableToConvertAttribute(attributeName, address, e);
+            }
+        }
+        return OptionalDouble.empty();
+    }
+
+    private ModelNode readAttributeValue(PathAddress address, String attributeName) {
+        final ModelNode readAttributeOp = new ModelNode();
+        readAttributeOp.get(OP).set(READ_ATTRIBUTE_OPERATION);
+        readAttributeOp.get(OP_ADDR).set(address.toModelNode());
+        readAttributeOp.get(ModelDescriptionConstants.INCLUDE_UNDEFINED_METRIC_VALUES).set(false);
+        readAttributeOp.get(NAME).set(attributeName);
+        ModelNode response = modelControllerClient.execute(readAttributeOp);
+        String error = getFailureDescription(response);
+        if (error != null) {
+            throw LOGGER.unableToReadAttribute(attributeName, address, error);
```

Comment:
The way getValue() is implemented it should be possible to check for WFLYCTL0216 and return an undefined ModelNode. I believe that would eliminate the https://issues.redhat.com/browse/WFLY-11933 issue for this metrics impl. With MP Metrics it's harder.

I expect you know all this which is why you use OptionalDouble. ;)